jina/serve/runtimes/gateway/http/app.py

```diff
@@ -227,6 +228,72 @@ async def _render_custom_swagger_html(req: Request) -> HTMLResponse:
 
         app.add_route(docs_url, _render_custom_swagger_html, include_in_schema=False)
 
+    if not args.no_graphql_endpoint:
+        with ImportExtensions(required=True):
+            from dataclasses import asdict
+
+            import strawberry
+            from docarray import DocumentArray
+            from strawberry.fastapi import GraphQLRouter
+            from docarray.document.strawberry_type import StrawberryDocument
+            from docarray.document.strawberry_type import (
+                JSONScalar,
+                StrawberryDocumentInput,
+            )
+
+            async def get_docs_from_endpoint(
+                data, target_executor, parameters, exec_endpoint
+            ):
+                req_generator_input = {
+                    'data': [asdict(d) for d in data],
+                    'target_executor': target_executor,
+                    'parameters': parameters,
+                    'exec_endpoint': exec_endpoint,
+                    'data_type': DataInputType.DICT,
+                }
+
+                if (
+                    req_generator_input['data'] is not None
+                    and 'docs' in req_generator_input['data']
+                ):
+                    req_generator_input['data'] = req_generator_input['data']['docs']
+
+                response = await _get_singleton_result(
+                    request_generator(**req_generator_input)
+                )
+                return DocumentArray.from_dict(response['data']).to_strawberry_type()
+
+            @strawberry.type
+            class Mutation:
+                @strawberry.mutation
+                async def docs(
+                    self,
+                    data: Optional[List[StrawberryDocumentInput]] = None,
```

Comment:
Shouldn't these be defined as a `StrawberryRequestModel` of type `input` & used here?

Reply:
Yes we could couple all inputs into one input type, but I see two issues with that:

1. From the user perspective, they now have to put all of their inputs into this type when defining their mutation. This creates unnecessary nesting. So instead of being able to do
```
mutation {docs(data: {...}, execEndpoint: "/foo", ...)
                ...            
                }
```
they would have to to something like this:
```
mutation {docs(inputType: {data: {...}, execEndpoint: "/foo", ...})
                ...            
                }
```

2. I think the above makes sense if you are grouping things together that are conceptually part of the same object, like in the strawberry example, the x an y coordinates of a point. But our inputs don't belong to the same object, so I believe that they should be separate and on the top level.

Reply:
Makes sense, then.